File: 韩/src/main/java/com/editor/core/TextEditor.java
package com.editor.core;

import com.editor.command.CommandInvoker;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 纯文本编辑器实现，内部使用按行存储的结构管理文本内容。
 */
public class TextEditor implements Editor {
    private String fileName;
    private final List<String> lines = new ArrayList<>();
    private boolean modified;
    private boolean logEnabled;
    private final CommandInvoker invoker = new CommandInvoker();

    /**
     * 创建文本编辑器，并根据初始内容决定是否自动开启日志。
     *
     * @param fileName 文件名
     * @param initialContent 初始内容
     */
    public TextEditor(String fileName, String initialContent) {
        this.fileName = fileName;
        loadContentFromString(initialContent);
        this.modified = false;
        this.logEnabled = shouldAutoEnableLog();
    }

    @Override
    public String getFileName() {
        return fileName;
    }

    @Override
    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    @Override
    public CommandInvoker getCommandInvoker() {
        return invoker;
    }

    @Override
    public boolean isModified() {
        return modified;
    }

    @Override
    public void markModified() {
        this.modified = true;
    }

    @Override
    public void markSaved() {
        this.modified = false;
    }

    @Override
    public boolean isLogEnabled() {
        return logEnabled;
    }

    @Override
    public void setLogEnabled(boolean enabled) {
        this.logEnabled = enabled;
    }

    @Override
    public String getContentForSave() {
        return String.join("\n", lines);
    }

    @Override
    public void loadContentFromString(String content) {
        lines.clear();
        if (content == null || content.isEmpty()) {
            return;
        }
        lines.addAll(Arrays.asList(content.split("\\R", -1)));
    }

    @Override
    public String getType() {
        return "text";
    }

    /**
     * @return 当前完整内容的字符串表示
     */
    public String getContentAsString() {
        return getContentForSave();
    }

    /**
     * 使用快照内容恢复编辑器状态。
     *
     * @param content 旧内容
     * @param wasModified 旧的修改标记
     */
    public void restoreFromString(String content, boolean wasModified) {
        loadContentFromString(content);
        if (wasModified) {
            markModified();
        } else {
            markSaved();
        }
    }

    /**
     * 在文件尾部追加一整行文本。
     *
     * @param text 追加内容
     */
    public void appendLine(String text) {
        if (text == null) {
            text = "";
        }
        lines.add(text);
        markModified();
    }

    /**
     * 在指定行列插入文本。
     * 当文本中包含换行符时，会将原有单行拆分成多行。
     *
     * @param line 行号，从 1 开始
     * @param col 列号，从 1 开始
     * @param text 待插入文本
     */
    public void insert(int line, int col, String text) {
        if (text == null) {
            text = "";
        }
        if (lines.isEmpty()) {
            if (line != 1 || col != 1) {
                throw new IllegalArgumentException("空文件只能在1:1位置插入");
            }
            loadInsertedIntoEmpty(text);
            markModified();
            return;
        }

        validatePosition(line, col);
        int lineIndex = line - 1;
        String currentLine = lines.get(lineIndex);
        String prefix = currentLine.substring(0, col - 1);
        String suffix = currentLine.substring(col - 1);
        String[] insertedLines = text.split("\\n", -1);

        // 多行插入时，首行接前缀，中间行为新增整行，末行接原后缀。
        if (insertedLines.length == 1) {
            lines.set(lineIndex, prefix + text + suffix);
        } else {
            lines.set(lineIndex, prefix + insertedLines[0]);
            for (int i = 1; i < insertedLines.length - 1; i++) {
                lines.add(lineIndex + i, insertedLines[i]);
            }
            lines.add(lineIndex + insertedLines.length - 1,
                    insertedLines[insertedLines.length - 1] + suffix);
        }
        markModified();
    }

    /**
     * 从指定位置开始删除固定长度的字符，删除范围不允许跨行。
     *
     * @param line 行号，从 1 开始
     * @param col 列号，从 1 开始
     * @param len 删除长度
     */
    public void delete(int line, int col, int len) {
        if (len < 0) {
            throw new IllegalArgumentException("删除长度不能为负数");
        }
        if (lines.isEmpty()) {
            throw new IllegalArgumentException("文件为空");
        }
        validatePosition(line, col);
        int lineIndex = line - 1;
        String currentLine = lines.get(lineIndex);
        int start = col - 1;
        int end = start + len;
        if (end > currentLine.length()) {
            throw new IllegalArgumentException("删除长度超出行尾");
        }
        lines.set(lineIndex, currentLine.substring(0, start) + currentLine.substring(end));
        markModified();
    }

    /**
     * 先删除再插入，从而实现替换操作。
     *
     * @param line 行号，从 1 开始
     * @param col 列号，从 1 开始
     * @param len 被替换长度
     * @param text 替换后的文本
     */
    public void replace(int line, int col, int len, String text) {
        delete(line, col, len);
        insert(line, col, text);
    }

    /**
     * 按行号范围生成展示文本。
     *
     * @param startLine 起始行号，可为 null
     * @param endLine 结束行号，可为 null
     * @return 带行号的展示字符串
     */
    public String show(Integer startLine, Integer endLine) {
        if (lines.isEmpty()) {
            return "";
        }
        int start = startLine == null ? 1 : startLine;
        int end = endLine == null ? lines.size() : endLine;
        if (start < 1 || end < start || end > lines.size()) {
            throw new IllegalArgumentException("显示范围越界");
        }
        StringBuilder builder = new StringBuilder();
        for (int i = start; i <= end; i++) {
            builder.append(i).append(": ").append(lines.get(i - 1));
            if (i < end) {
                builder.append('\n');
            }
        }
        return builder.toString();
    }

    /**
     * @return 当前文本总行数
     */
    public int getLineCount() {
        return lines.size();
    }

    /**
     * 处理空文件上的首次插入。
     *
     * @param text 要插入的文本
     */
    private void loadInsertedIntoEmpty(String text) {
        if (text.isEmpty()) {
            lines.add("");
            return;
        }
        lines.addAll(Arrays.asList(text.split("\\n", -1)));
    }

    /**
     * 校验插入、删除、替换使用的行列坐标是否合法。
     *
     * @param line 行号
     * @param col 列号
     */
    private void validatePosition(int line, int col) {
        if (line < 1 || line > lines.size()) {
            throw new IllegalArgumentException("行号或列号越界");
        }
        String target = lines.get(line - 1);
        if (col < 1 || col > target.length() + 1) {
            throw new IllegalArgumentException("行号或列号越界");
        }
    }

    /**
     * 当第一行是 # log 时，自动将日志开关初始化为开启状态。
     *
     * @return 是否应自动开启日志
     */
    private boolean shouldAutoEnableLog() {
        return !lines.isEmpty() && "# log".equals(lines.get(0));
    }
}

File: 侯/editor/editor/text_editor.py
# editor/editor/text_editor.py
from typing import List, Optional
from .base import Editor
from editor.command.history import CommandHistory
from editor.persistence.file import load_file, save_file

class TextEditor(Editor):
    """文本编辑器实现"""
    def __init__(self, file_path: str):
        self._file_path = file_path
        self._modified = False
        self._lines: List[str] = []
        self._command_history = CommandHistory()
        self._log_enabled = False

    # ===== 核心编辑方法 =====
    def append(self, text: str) -> None:
        """追加文本到末尾"""
        self._lines.append(text)
        self._modified = True

    def insert(self, line_num: int, col_num: int, text: str) -> None:
        """在指定位置插入文本（line:col 从1开始）"""
        if not self._lines:
            if line_num != 1 or col_num != 1:
                raise Exception("空文件只能在1:1位置插入")
            self._lines.append(text)
            self._modified = True
            return

        if line_num < 1 or line_num > len(self._lines):
            raise Exception("行号越界")
        target_line = self._lines[line_num - 1]
        if col_num < 1 or col_num > len(target_line) + 1:
            raise Exception("列号越界")

        # 处理包含换行符的文本，按行分割
        split_text = text.split("\n")
        if len(split_text) == 1:
            new_line = target_line[:col_num-1] + text + target_line[col_num-1:]
            self._lines[line_num - 1] = new_line
        else:
            # 多行插入
            first_part = target_line[:col_num-1] + split_text[0]
            self._lines[line_num - 1] = first_part
            for i in range(1, len(split_text)-1):
                self._lines.insert(line_num + i - 1, split_text[i])
            last_part = split_text[-1] + target_line[col_num-1:]
            self._lines.insert(line_num + len(split_text) - 2, last_part)
        self._modified = True

    def delete(self, line_num: int, col_num: int, length: int) -> None:
        """删除指定位置开始的 length 个字符（不可跨行）"""
        if not self._lines:
            raise Exception("文件为空，无法删除")
        if line_num < 1 or line_num > len(self._lines):
            raise Exception("行号越界")
        target_line = self._lines[line_num - 1]
        if col_num < 1 or col_num > len(target_line):
            raise Exception("列号越界")
        if col_num + length - 1 > len(target_line):
            raise Exception("删除长度超出行尾")

        new_line = target_line[:col_num-1] + target_line[col_num-1+length:]
        self._lines[line_num - 1] = new_line
        self._modified = True

    def replace(self, line_num: int, col_num: int, length: int, text: str) -> None:
        """替换指定位置的字符（先删除后插入）"""
        self.delete(line_num, col_num, length)
        self.insert(line_num, col_num, text)

    def show(self, start_line: Optional[int] = None, end_line: Optional[int] = None) -> None:
        """显示指定范围的文本（按行）"""
        if not self._lines:
            print("文件为空")
            return

        start = start_line or 1
        end = end_line or len(self._lines)
        start = max(1, start)
        end = min(len(self._lines), end)

        if start > end:
            print("起始行号大于结束行号")
            return

        for i in range(start-1, end):
            print(f"{i+1}: {self._lines[i]}")

    # ===== Editor 接口实现 =====
    def get_file_path(self) -> str:
        return self._file_path

    def set_file_path(self, path: str) -> None:
        self._file_path = path

    def is_modified(self) -> bool:
        return self._modified

    def set_modified(self, modified: bool) -> None:
        self._modified = modified

    def get_command_history(self) -> CommandHistory:
        return self._command_history

    def load_content(self) -> None:
        from editor.persistence.file import load_file
        self._lines = load_file(self._file_path)
    # 文件第一行是# log，自动启用日志
        if self._lines and self._lines[0].strip() == "# log":
            self._log_enabled = True
        self._modified = False

    def save_content(self) -> None:
        save_file(self._file_path, self._lines)
        self._modified = False

    def close(self) -> None:
        self._lines.clear()
        self._command_history.clear()

    def get_lines(self) -> List[str]:
        return self._lines.copy()  # 返回副本，避免外部修改

    def is_log_enabled(self) -> bool:
        return self._log_enabled

    def set_log_enabled(self, enabled: bool) -> None:
        self._log_enabled = enabled
    
    def set_lines(self, lines: list[str]) -> None:
        """从快照恢复文本内容"""
        self._lines = lines.copy()
        self._modified = True